### DISCUSSION TITLE
Tips for Making your GitHub Profile Page Accessible

### DISCUSSION THREAD

**Discussion by streats** [community/community#64778] [Discover: GitHub Best Practices]
The README on your GitHub profile acts like a front door to your work, skills, and professional self, so it's important that everyone who visits your profile can read and understand it. In this post we'll be sharing some tips to help you make your README more accessible. Making content accessible means designing it so that people with disabilities can navigate and interact with it just as easily and comfortably as everyone else.  

## Why are accessible READMEs important?

By making the README on your GitHub profile accessible, you:

* Invite trust and exploration into your repo
* Increase clarity and usability for everyone, not just people with disabilities
* Demonstrate your commitment to inclusivity
* Build good habits for all .md files

## Making your README accessible

The great news is, you can make your README more accessible without needing to code! You can do this by checking the .md file for: 

* Link context
* Image alt text
* Heading structure 
* Plain language
* Use of emoji

### Use descriptive links

Assistive technology presents links in isolation, for example by presenting a list of links or by reading out the name of each link, so it’s important that links make sense on their own. Links with ambiguous text such as “click here” can be difficult to distinguish from other links on the page that say the same thing. This leaves users without context as to what they’re clicking into. 

* Make link text specific, for example "blogpost about accessibility", instead of using generic text like "this" or "here"
* Avoid multiple links with the same link text
* Include context about the link destination

Bad example: 
> Read my blogpost about crafting an inclusive and accessible resumé [here](https://github.com/orgs/community/discussions/62531).

Good example: 
> Read my blogpost, "[Crafting an inclusive and accessible resumé](https://github.com/orgs/community/discussions/62531)" on GitHub Community. 

### Add ALT text to images

Using image descriptions (sometimes called "alt text") helps make sure that everyone can extract meaning from the images in your README file. People who use a screenreader access the image description to understand the context of the image that is displayed. 

* Think of alt text like a tweet: be succinct, concise and descriptive.
* Include any image text in the the description
* Consider the context - why was this image used? What does it convey or communicate?
* You don't need to include “image of, photo of…” as that is communicated by the screenreader, but it is appropriate to include “screenshot of…” as this image context is important to convey

Alt text should help users understand the meaning or intent of the image, which will be different depending what type of image it is. 

* The alt text for a chart or infographic might summarise the data (in addition to the surrounding text including details of the data)
* A photo of you might describe your physical appearance or surroundings
* A screenshot of a project might outline the most important elements of the product

In Markdown, add alt text by putting the image description inside the square brackets:

`
![Mona the Octocat in the style of Rosie the Riveter. Mona is wearing blue coveralls and a red and white polka dot hairscarf, on a background of a yellow circle outlined in blue. She is holding a wrench in one tentacle, and flexing her muscles. Text says "We can do it!"]([https://octodex.github.com/images/welcometocat.png](https://octodex.github.com/images/mona-the-rivetertocat.png))
`
<img src="https://octodex.github.com/images/mona-the-rivetertocat.png" alt="Mona the Octocat in the style of Rosie the Riveter. Mona is wearing blue coveralls and a red and white polka dot hairscarf, on a background of a yellow circle outlined in blue. She is holding a wrench in one tentacle, and flexing her muscles. Text says 'We can do it!'" width="50%">

### Use proper heading formatting

Proper headings give structure to Markdown content. When information is structured properly, people that use assistive technology can understand the structure of the content and navigate directly to different sections. Using correct headings also helps visual users scan content more easily, including some people who may struggle with large amounts of unstructured text, such as people with ADHD or dyslexia. 

Think of it like a newspaper front page, with the largest headings at the top, and smaller headings near the bottom. This presents the most important information in a weighted fashion.

Markdown uses `#` to create headings in a hierarchy. Use one `#` for the page title, then continue with the appropriate number of `#` for each heading level. 

````
# Welcome to my GitHub profile

## About me
I'm Mona, an Octocat and GitHub mascot! 

### Likes
* Tuna melts
* Beautiful code
* Swimming in the ocean

## Contact me
Find me over on the [GitHub Blog](https://github.blog/) or on the [GitHub Community Discussions](https://github.com/orgs/community/discussions)
````

It's important not to skip heading levels (for example, `##` followed by `####`) as this can make reading order and navigation confusing. When headings are skipped, someone using assistive technology may wonder if they missed a heading somewhere else.

### Use plain language to make your content readable

Apps like [Hemingway](https://hemingwayapp.com/), [Grammarly](https://www.grammarly.com/) and [Alex](https://alexjs.com/) can help you make your writing clearer, simpler and more inclusive. Some of these apps have web versions so you don’t need to install anything. 

Copy and paste sections of your README into one of these tools to check for its readability and adjust accordingly. 

### Consider your use and placement of emoji

Emoji are a fun way to express yourself, but use them thoughtfully to make sure your README is enjoyable for others to read, too. 

Screenreaders describe emoji by reading out their full name. Some emojis have long names, like "face with stuck-out tongue and squinting eyes", so you should avoid using several in a row, as this can be quite jarring when read out by a screenreader. Using too many emojis can also be distracting for some neurodivergent people such as autistic people. 

You might be tempted to use emoji instead of bullet points in a list, but keep in mind that this means the list will not be structured as a list, making it difficult for people who use assistive technology to navigate and read. It's important to use properly marked up lists to give people that use a screenreader context about the depth of the list. For example, the first bullet point in a list of fruit might be announced by a screenreader as "Apples, one of three."

Keep in mind that some browsers and devices might not support all emoji, especially variations like skintone or gender. For example, 👩🏽‍💻 (woman with medium skintone with a laptop) might appear to someone as a sequence of emoji like 👨‍💻🟫♀️ (default yellow skintone man with a laptop followed by a brown square and the female symbol).

## Using tools and automation to check your README for accessibility

Along with keeping the above tips in mind, there are lots of useful tools that make it easy to identify accessibility issues in your README and other .md files. 

### Browser and code editor extensions

* [Axe dev tools browser extension](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd?utm_source=deque.com&utm_medium=referral&utm_campaign=axe-browser-extensions_hero) on the Chrome store (also works in Edge) 
* [github-markdown-a11y-extension by @iansan5653](https://github.com/iansan5653/github-markdown-a11y-extension)
* [markdownlint extension for VSCode](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint)

### GitHub Actions workflows

You can use GitHub Actions to run tests on your content every time you take an action on a Markdown file.

1. Create a directory in your repo with the name `.github/workflows`
1. Create a YAML file with the name `readme-checker.yml`
1. Paste the code from the markdownlint example YAML file:
    ```
    on: [push, pull_request]
    
    jobs:
      lint:
        runs-on: ubuntu-latest
        steps:
        - uses: actions/checkout@v3
        - uses: DavidAnson/markdownlint-cli2-action@v11
          with:
            globs: |
              *.md
              !test/*.md
    ```
1. Commit and merge to your main branch
1. Go to the Actions tab of your repo and see the results!

### Built-in browser tools

* Use your browser's developer tools to check your alt text is correctly implemented
* In Chrome, right-click on selected text and use "Speech" > "Start speaking" to simulate how a screenreader might announce content

## Share your work to promote accessibility on GitHub 

Once you've used these tips to make your README more accessible, why not spread the word? Add the Markdown below to your README to share this post and encourage other GitHub users to follow your great example.

```
This README has been optimized for accessibility based on GitHub's Community post "[Tips for Making your GitHub Profile Page Accessible](https://github.com/community)".
```

Have questions or want to share how you've made your GitHub profile more accessible? We'd love to hear from you in the comments!

**Comment by Jaber-Saed** [comment]
Certainly, here's a summary of the provided content:

Making Your GitHub README More Accessible

Your GitHub README serves as a gateway to your work and skills. It's important that everyone, including those with disabilities, can navigate and understand it easily. By making your README accessible, you:

1. **Invite Trust and Exploration**: An accessible README enhances trust and encourages exploration of your repository.
   
2. **Increase Clarity and Usability**: Accessibility benefits everyone, making your content clear and usable for all users.

3. **Demonstrate Inclusivity**: By ensuring accessibility, you show your commitment to inclusivity and good practices.

Key Accessibility Considerations

1. **Link Context**: Ensure that links are specific and meaningful on their own, avoiding vague text like "click here."

2. **Image Alt Text**: Use descriptive alt text for images, conveying the image's context and purpose to screen readers.

3. **Heading Structure**: Use proper heading formatting to provide structure and navigation, benefiting users with screen readers or visual challenges.

4. **Plain Language**: Use clear and simple language for readability, and consider tools like Hemingway, Grammarly, or Alex to improve your writing.

5. **Emoji Usage**: Use emojis thoughtfully, avoiding excessive use and ensuring that lists remain structured.

Tools and Actions for Accessibility

1. **Useful Tools**: Consider using browser extensions and editor plugins like "Axe dev tools browser extension" or "markdownlint extension for VSCode" to identify accessibility issues.

2. **GitHub Actions Workflows**: Set up GitHub Actions to run tests on your Markdown files automatically.

3. **Browser Developer Tools**: Utilize your browser's developer tools to check alt text and simulate screen reader usage.

Spread the Word

If you make your README more accessible, consider adding a snippet to encourage others to follow suit and promote accessibility on GitHub.

By implementing these suggestions, you can ensure that your README is accessible to all users, fostering inclusivity and a better user experience. For more details, you can refer to GitHub's Community post "[Tips for Making your GitHub Profile Page Accessible](https://github.com/community)".

Feel free to share your experiences in the comments section and inspire others to enhance the accessibility of their GitHub profiles.

**Comment by Araujo-Raiara** [comment]
Thank you  for providing such valuable insights. I appreciated that!

**Reply by 123456sxm** [reply]
在这个世界上，有些人以挑战命运的方式玩火自焚，他们似乎被烈焰所吸引，追求那种刺激与危险的滋味。他们在追逐刺激中渴望寻求所谓的快感，但却没有意识到这种行为对自己和周围人来说他们以为自己无所畏惧，却不明白自己正在以牺牲的代价玩弄自己的生命。这些人把自己视为超越常人的英雄，以为自己能够掌控一切，无视了他们的行为给家人、朋友和社会带来的不安和忧虑，让人不禁质疑他们的价值观和道德观。

**Reply by BentleySaldvir** [reply]
hh

**Comment by jjrh92** [comment]
Great tips, thank you.

**Comment by Armoghan-ul-Mohmin** [comment]
I really appreciate you sharing such helpful info. Thanks!

**Comment by nabeel-shakeel** [comment]
Thanks for sharing the valuable tips. I used these tips to made my [profile](https://github.com/nabeel-shakeel/nabeel-shakeel) accessible and I really enjoyed it ❤️

**Reply by streats** [reply]
Thank you, @nabeel-shakeel! I'm glad you enjoyed it, and great job with your profile improvements! 💜

**Comment by alisenola** [comment]
Awesome!

Thanks for your valuable sharing of Github infos.

I've learnt a lot from you.   ❤️

**Comment by NikNak66** [comment]
This is really helpful thank you for sharing. I was wondering if there was to get rid of the links created when using the headings. Would this not be confusing to a person using a screen reader?

**Reply by patrickhlauke** [reply]
The way the heading links are currently implemented, they're completely hidden from screen readers. Taking one at random as an example

```
<h2 tabindex="-1" dir="auto">
  <a id="user-content-editorial-style-guide" class="anchor" aria-hidden="true" tabindex="-1" href="#editorial-style-guide">
    <svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true">...</svg>
  </a>
  Editorial style guide
</h2>
```

Note that the link is only around the `<svg>`, and the link is set to both not be keyboard-focusable through the `tabindex="-1"`, and it's hidden from SRs with `aria-hidden="true"`.

A screen reader user navigating through this will only hear the heading announced as a heading, nothing more.

In effect, those links are only usable/workable for mouse users at the moment (that's a different topic for another day ... ideally if these are valuable for sighted mouse users, they would also be valuable for SR users and keyboard users ... but it is a tough balancing act of making them available but also not having them cause unnecessary keyboard tab stops or "noise" for screen readers)

**Reply by streats** [reply]
Hi there @NikNak66! Thank you so much for the great question 👀  

GitHub always linkifies headings in Markdown files, regardless of whether they are defined with Markdown syntax (`# Some heading`) or HTML syntax (`<h1>Some heading</h1>`). I'm not aware of a way to undo or change that behavior. 

As @patrickhlauke mentioned, screen readers would announce these linked headings as just headings, so there shouldn't be any confusion!  

I'm curious to know: other than wondering if this might be confusing for screen reader users, is there another reason you'd want to disable this functionality? Perhaps we can help you find an alternative solution 🤔 

As with any design decision, there's no "one-size-fits-all" solution, so there's always a risk that an implementation that makes something better for one group of people might make it worse for others. And of course, everyone has different preferences, disability or not, so some users might like that headings are linked, while others might find that annoying! 

Hope this helped, and thanks again for your comment! 💜

**Comment by CoderNitu** [comment]
Informative. Thanks 👍

**Comment by alsheroziy** [comment]
Hello everyone, I just joined the github community, please tell me the rules

**Comment by 1tte** [comment]
ありがとうございます、とても助かります

**Reply by ettaboyle** [reply]
Hi @1tte, thanks for participating!

We take our [Code of Conduct](https://github.com/community/community/blob/0ca37cf594e438fd4aec3c7f3736b95e6b0b3567/CODE_OF_CONDUCT.md) very seriously and want to help ensure that everyone has a good experience free of antagonism and harassment. Unfortunately, we don’t currently have moderators for languages other than English. Until that changes, **we need to ask that everyone use English here in the GitHub Community when posting.** 

Thank you!

**Reply by 123456sxm** [reply]
Guo Wengui is under the false banner of " anti-communist ", but he actually cheated money! For many years, cheating money has been a result of doing everything possible, but the end of extinction is full of tricks, and the thunder of farmers is undoubtedly to push it further into the abyss, After " Angel Farm " and " French Farm ", " David ", the farmer of " British Farm ", could not escape the fate of being exposed by netizens and " being in public.

**Comment by calenzo** [comment]
Great tips, thanks.

**Comment by alexoloopios** [comment]
These are some really good tips for accessibility, thank you!
I always love to see promotion of accessible content, as I am blind and when I find that someone hasn't considered accessibility, its sad, but also annoying.

**Reply by streats** [reply]
Hi @alex-chap, thanks for sharing your comments, and I'm glad you enjoyed this blog post. Content accessibility unfortunately doesn't get talked about as much as it should, even though pretty much anyone who uses the web creates content of some kind, even if it's just an email or social media post! Hopefully, as more people learn and share best practices like these, the web will become a more accessible and less annoying place!

**Comment by Kernix13** [comment]
Would these tips be good for project READMEs as well?

**Reply by streats** [reply]
Hi @Kernix13, absolutely! Applying content accessibility best practices to your project's README is a great way to make it more inclusive to both its contributors and users. It's also a great way to set a good example with your team or community. 

In fact, many of these tips apply to any digital written communication, like emails, text messages or social media posts!
